### ISSUE TITLE
Balls can escape arena - clamp velocities with custom behavior.

### ISSUE THREAD

**Issue by ryant26** [polyball/polyball#150]
- They need to be velocity damped in some way

**Comment by kdbanman** [comment]
I was hoping we could listen to a post-impulse-response channel and clamp the resulting velocities, but that doesn't seem to exist.  I'll keep researching.

The big hammer fix here is to listen to the `integrate:positions` channel and limit the velocities of balls after every integration.  This is the least performant option and will mess with balls as they interact with black hole attractors, though.

**Comment by ryant26** [comment]
Yea I was thinking we would have to do the velocities at every position update. Running through like 10-15 balls hopefully shouldn't impact performance too badly.... Let me know if you find a better solution

**Comment by kdbanman** [comment]
I have a solution that's more complex than running on every position update, but it should be much more efficient and won't mess with black holes.

We only care about when balls and paddles interact for velocity clamping.  We can know when that's happening by listening to the `collisions:detected` channel.  That happens *before* impulse responses are applied, though, so we would need to clamp paddle velocities there, which is going to mess with how the paddles feel.

However:

```
on collisions detected:
    if bodies are paddle and ball:
        mark ball for velocity clamp on next integration

before integration:
    for each marked ball:
        clamp velocity magnitude to some maximum
        unmark ball
```

That latter algorithm still runs on every integration, but it almost always does nothing.  It only affects balls that were touched by a paddle, and it only affects those balls for a single integration.

**Comment by kdbanman** [comment]
Also the balls will be less prone to flying through the arena if we make the arena walls thicker... :P

**Comment by ryant26** [comment]
Yea, I still think that you can currently hit balls so hard that no one would ever have a chance of defending your shot.

**Comment by ryant26** [comment]
>That happens before impulse responses are applied, though, so we would need to clamp paddle velocities there, which is going to mess with how the paddles feel.

Why would we have to clamp paddle velocities?

**Comment by kdbanman** [comment]
Sorry that paragraph is pretty unclear.  Here's a clearer version:

We want to limit ball velocity from paddle impacts.  There are two choices:

1. Limit paddle velocity before impact.
2. Limit ball velocity after impact.

We're already kinda doing option 1, but it has issues.  I tried polyball on a newer macbook pro that throws mousemove events less often, and the paddles moved really, really sluggishly.  Whether it's from the hardware or from aggregation interval changes, I feel like frequency of move commands should not change how responsive the paddles are.

The pseudocode above implements option 2.  It means that the paddle can be as fast and responsive as we like without affecting ball-paddle impacts.  (I *swear* I got a message from you that says this already, but I can't find it now.  Maybe I was dreaming.)
